security/pkg/nodeagent/sds/sdsservice.go

```diff
@@ -504,13 +510,20 @@ func pushSDS(con *sdsConnection) error {
 	conID := con.conID
 	secret := con.secret
 	resourceName := con.ResourceName
+	sdsPushTime := con.sdsPushTime
 	con.mutex.RUnlock()
 
+	conIDresourceNamePrefix := sdsLogPrefix(conID, resourceName)
```

Comment:
for workload case, if token is the same; will the cert rotation still work with this change ? (sdsconnection keep open, push timestamp is set after 1st rotation, which prevents later push)

Reply:
This change does not affect token. If there is a SDS request received, and it is an ACK. We reset sdsPushTime to zero, SecretExist() returns true, and continue. Later when we have new secret to push, we call pushSDS, and because sdsPushTime is reset to zero, we can push the new secret and set sdsPushTime to a non-zero value. The sdsPushTime is reset on every new SDS request, and then set after each SDS response is pushed.

Reply:
could below event sequence may happen ? (though in theory should be rare)
1. ack response arrive first after below ```con.stream.Send```
2. ```con.sdsPushTime = time.Now()``` is called

Reply:
By looking at Envoy code, I think it always processes response first and then send next SDS request. https://github.com/envoyproxy/envoy/blob/27b41b888d5ba624f2a6098c8b9f395e8ef95cbf/source/common/config/grpc_mux_impl.cc#L202.
I think the thread model is, a new connection comes in--> create goroutine to run StreamSecrets-->pushSDS-->wait for next event at StreamSecrets. If pushSDS is not returned, StreamSecrets on that goroutine does not have chance to handle next event.

We can lock the entire method to prevent unexpected SDS request during pushSDS. Let me know which way you would prefer.

Reply:
my understanding is pushsds and handing coming request could happened simultaneously in different go-routine; we could test by sleeping between ```con.stream.Send``` and return. I'm ok to proceed current change as it is.

Reply:
I have updated the pushSDS method and used lock to protect the sdsPushTime until push is done.